csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs

```diff
@@ -924,6 +924,17 @@ private void DownloadMissingPackages(List<FileInfo> allFiles, ISet<string> dllPa
                 return;
             }
 
+            var multipleVersions = notYetDownloadedPackages
+                .GroupBy(p => p.Name)
+                .Where(g => g.Count() > 1)
+                .Select(g => g.Key);
+
+            foreach (var package in multipleVersions)
+            {
+                logger.LogWarning($"Found multiple not yet restored packages with name '{package}'.");
+                notYetDownloadedPackages.Remove(new(package, PackageReferenceSource.PackagesConfig));
```

Comment:
Isn't it problematic to alter the collection that you are iterating over? For example, running
```csharp
List<int> a = [1, 2, 3, 4, 5];

var strings = a.Select(x => x.ToString());

foreach (var s in strings)
{
    a.Remove(4);
    Console.WriteLine(s);
}
```
results in
```
Unhandled exception. System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
```

Reply:
It seems to be working, but I'm adding a `.ToList()` to the end of the calls to be safe.

Reply:
I think this is because the `GroupBy` implementation creates a "lookup" on the first call to `MoveNext` which does a full enumeration of the source. That is, the enumeration has already taken place before the body of the `foreach` loop is executed the first time (this is contrary to the `Select` example which yield returns the result of applying the function to the first element).